File: core/src/main/java/com/uiptv/service/SeriesWatchingNowSnapshotService.java
package com.uiptv.service;

import com.uiptv.db.SeriesWatchingNowSnapshotDb;
import com.uiptv.model.Account;
import com.uiptv.model.Channel;
import com.uiptv.model.SeriesWatchingNowSnapshot;
import com.uiptv.shared.Episode;
import com.uiptv.shared.EpisodeList;
import org.json.JSONArray;

import java.util.ArrayList;
import java.util.List;

import static com.uiptv.util.StringUtils.isBlank;

@SuppressWarnings("java:S6548")
public class SeriesWatchingNowSnapshotService {
    private SeriesWatchingNowSnapshotService() {
    }

    private static class SingletonHelper {
        private static final SeriesWatchingNowSnapshotService INSTANCE = new SeriesWatchingNowSnapshotService();
    }

    public static SeriesWatchingNowSnapshotService getInstance() {
        return SingletonHelper.INSTANCE;
    }

    public SeriesWatchingNowSnapshot getSnapshot(String accountId, String categoryId, String seriesId) {
        if (isBlank(accountId) || isBlank(seriesId)) {
            return null;
        }
        String normalizedCategory = normalize(categoryId);
        List<String> candidateIds = buildSeriesIdCandidates(seriesId);
        SeriesWatchingNowSnapshot exact = findExactSnapshot(accountId, normalizedCategory, candidateIds);
        return exact != null ? exact : findLatestSnapshot(accountId, candidateIds);
    }

    public EpisodeList loadEpisodeList(String accountId, String categoryId, String seriesId) {
        SeriesWatchingNowSnapshot snapshot = getSnapshot(accountId, categoryId, seriesId);
        if (snapshot == null || isBlank(snapshot.getEpisodesJson())) {
            return new EpisodeList();
        }
        JSONArray payload;
        try {
            payload = new JSONArray(snapshot.getEpisodesJson());
        } catch (Exception _) {
            return new EpisodeList();
        }
        EpisodeList list = new EpisodeList();
        for (int i = 0; i < payload.length(); i++) {
            Channel channel = readChannel(payload, i);
            if (channel != null) {
                list.getEpisodes().add(toEpisode(channel));
            }
        }
        return list;
    }

    public List<Channel> loadChannels(String accountId, String categoryId, String seriesId) {
        SeriesWatchingNowSnapshot snapshot = getSnapshot(accountId, categoryId, seriesId);
        if (snapshot == null || isBlank(snapshot.getEpisodesJson())) {
            return List.of();
        }
        JSONArray payload;
        try {
            payload = new JSONArray(snapshot.getEpisodesJson());
        } catch (Exception _) {
            return List.of();
        }
        List<Channel> channels = new ArrayList<>();
        for (int i = 0; i < payload.length(); i++) {
            Channel channel = readChannel(payload, i);
            if (channel != null) {
                channels.add(channel);
            }
        }
        return channels;
    }

    public void save(Account account,
                     String categoryId,
                     String seriesId,
                     String categoryDbId,
                     String seriesTitle,
                     String seriesPoster,
                     EpisodeList episodeList) {
        if (account == null || isBlank(account.getDbId()) || isBlank(seriesId) || episodeList == null
                || episodeList.getEpisodes() == null || episodeList.getEpisodes().isEmpty()) {
            return;
        }
        JSONArray episodesPayload = new JSONArray();
        for (Episode episode : episodeList.getEpisodes()) {
            Channel channel = toChannel(episode);
            if (channel != null) {
                episodesPayload.put(channel.toJson());
            }
        }
        if (episodesPayload.isEmpty()) {
            return;
        }
        SeriesWatchingNowSnapshot snapshot = new SeriesWatchingNowSnapshot();
        snapshot.setAccountId(account.getDbId());
        snapshot.setCategoryId(normalize(categoryId));
        snapshot.setSeriesId(canonicalizeSeriesId(seriesId));
        snapshot.setCategoryDbId(normalize(categoryDbId));
        snapshot.setSeriesTitle(normalize(seriesTitle));
        snapshot.setSeriesPoster(normalize(seriesPoster));
        snapshot.setEpisodesJson(episodesPayload.toString());
        snapshot.setUpdatedAt(System.currentTimeMillis());
        SeriesWatchingNowSnapshotDb.get().upsert(snapshot);
    }

    public void saveChannels(Account account,
                             String categoryId,
                             String seriesId,
                             String categoryDbId,
                             String seriesTitle,
                             String seriesPoster,
                             List<Channel> channels) {
        if (channels == null || channels.isEmpty()) {
            return;
        }
        EpisodeList list = new EpisodeList();
        for (Channel channel : channels) {
            if (channel != null) {
                list.getEpisodes().add(toEpisode(channel));
            }
        }
        save(account, categoryId, seriesId, categoryDbId, seriesTitle, seriesPoster, list);
    }

    public void clear(String accountId, String categoryId, String seriesId) {
        if (isBlank(accountId) || isBlank(seriesId)) {
            return;
        }
        String canonicalSeriesId = canonicalizeSeriesId(seriesId);
        String normalizedCategory = normalize(categoryId);
        SeriesWatchingNowSnapshotDb.get().clear(accountId, normalizedCategory, canonicalSeriesId);
        for (String candidateId : buildSeriesIdCandidates(seriesId)) {
            for (SeriesWatchingNowSnapshot candidate : SeriesWatchingNowSnapshotDb.get().getBySeries(accountId, candidateId)) {
                if (candidate != null) {
                    SeriesWatchingNowSnapshotDb.get().clear(accountId, normalize(candidate.getCategoryId()), normalize(candidate.getSeriesId()));
                }
            }
        }
    }

    public void clearAll() {
        SeriesWatchingNowSnapshotDb.get().clearAll();
    }

    private Channel toChannel(Episode episode) {
        if (episode == null || isBlank(episode.getId())) {
            return null;
        }
        Channel channel = new Channel();
        channel.setChannelId(episode.getId());
        channel.setName(episode.getTitle());
        channel.setCmd(episode.getCmd());
        channel.setSeason(episode.getSeason());
        channel.setEpisodeNum(episode.getEpisodeNum());
        channel.setExtraJson(episode.toJson());
        if (episode.getInfo() != null) {
            channel.setLogo(episode.getInfo().getMovieImage());
            channel.setDescription(episode.getInfo().getPlot());
            channel.setReleaseDate(episode.getInfo().getReleaseDate());
            channel.setRating(episode.getInfo().getRating());
            channel.setDuration(episode.getInfo().getDuration());
        }
        return channel;
    }

    private Episode toEpisode(Channel channel) {
        Episode parsed = Episode.fromJson(channel.getExtraJson());
        if (parsed != null && !isBlank(parsed.getId())) {
            hydrateParsedEpisode(parsed, channel);
            return parsed;
        }
        Episode episode = new Episode();
        episode.setId(channel.getChannelId());
        episode.setTitle(channel.getName());
        episode.setCmd(channel.getCmd());
        episode.setSeason(channel.getSeason());
        episode.setEpisodeNum(channel.getEpisodeNum());
        com.uiptv.shared.EpisodeInfo info = new com.uiptv.shared.EpisodeInfo();
        info.setMovieImage(channel.getLogo());
        info.setPlot(channel.getDescription());
        info.setReleaseDate(channel.getReleaseDate());
        info.setRating(channel.getRating());
        info.setDuration(channel.getDuration());
        episode.setInfo(info);
        return episode;
    }

    private String normalize(String value) {
        return value == null ? "" : value.trim();
    }

    private SeriesWatchingNowSnapshot findExactSnapshot(String accountId, String normalizedCategory, List<String> candidateIds) {
        for (String candidateId : candidateIds) {
            SeriesWatchingNowSnapshot match = SeriesWatchingNowSnapshotDb.get().getBySeries(accountId, normalizedCategory, candidateId);
            if (match != null) {
                return match;
            }
        }
        return null;
    }

    private SeriesWatchingNowSnapshot findLatestSnapshot(String accountId, List<String> candidateIds) {
        SeriesWatchingNowSnapshot latest = null;
        for (String candidateId : candidateIds) {
            latest = newerSnapshot(latest, SeriesWatchingNowSnapshotDb.get().getBySeries(accountId, candidateId));
        }
        return latest;
    }

    private SeriesWatchingNowSnapshot newerSnapshot(SeriesWatchingNowSnapshot current, List<SeriesWatchingNowSnapshot> candidates) {
        SeriesWatchingNowSnapshot latest = current;
        for (SeriesWatchingNowSnapshot candidate : candidates) {
            if (candidate != null && (latest == null || candidate.getUpdatedAt() > latest.getUpdatedAt())) {
                latest = candidate;
            }
        }
        return latest;
    }

    private Channel readChannel(JSONArray payload, int index) {
        Object value = payload.opt(index);
        return value == null ? null : Channel.fromJson(String.valueOf(value));
    }

    private void hydrateParsedEpisode(Episode parsed, Channel channel) {
        if (isBlank(parsed.getSeason())) {
            parsed.setSeason(channel.getSeason());
        }
        if (isBlank(parsed.getEpisodeNum())) {
            parsed.setEpisodeNum(channel.getEpisodeNum());
        }
        if (parsed.getInfo() == null) {
            parsed.setInfo(new com.uiptv.shared.EpisodeInfo());
        }
        mergeEpisodeInfo(parsed.getInfo(), channel);
    }

    private void mergeEpisodeInfo(com.uiptv.shared.EpisodeInfo info, Channel channel) {
        if (isBlank(info.getMovieImage())) {
            info.setMovieImage(channel.getLogo());
        }
        if (isBlank(info.getPlot())) {
            info.setPlot(channel.getDescription());
        }
        if (isBlank(info.getReleaseDate())) {
            info.setReleaseDate(channel.getReleaseDate());
        }
        if (isBlank(info.getRating())) {
            info.setRating(channel.getRating());
        }
        if (isBlank(info.getDuration())) {
            info.setDuration(channel.getDuration());
        }
    }

    private String canonicalizeSeriesId(String seriesId) {
        String raw = normalize(seriesId);
        if (isBlank(raw) || !raw.contains(":")) {
            return raw;
        }
        String[] parts = raw.split(":");
        String last = "";
        for (int i = parts.length - 1; i >= 0; i--) {
            String part = normalize(parts[i]);
            if (!isBlank(part)) {
                last = part;
                break;
            }
        }
        return isBlank(last) ? raw : last;
    }

    private List<String> buildSeriesIdCandidates(String seriesId) {
        String raw = normalize(seriesId);
        if (isBlank(raw)) {
            return List.of();
        }
        java.util.LinkedHashSet<String> candidates = new java.util.LinkedHashSet<>();
        candidates.add(canonicalizeSeriesId(raw));
        if (raw.contains(":")) {
            for (String part : raw.split(":")) {
                String normalized = normalize(part);
                if (!isBlank(normalized)) {
                    candidates.add(normalized);
                }
            }
        } else {
            candidates.add(raw + ":" + raw);
        }
        candidates.add(raw);
        return new ArrayList<>(candidates);
    }
}

File: server/src/main/kotlin/com/uiptv/service/SeriesWatchingNowSnapshotService.kt
package com.uiptv.service

import com.uiptv.db.SeriesWatchingNowSnapshotDb
import com.uiptv.model.Account
import com.uiptv.model.Channel
import com.uiptv.model.SeriesWatchingNowSnapshot
import com.uiptv.shared.Episode
import com.uiptv.shared.EpisodeInfo
import com.uiptv.shared.EpisodeList
import com.uiptv.util.StringUtils
import com.uiptv.util.json.parseJsonArray
import com.uiptv.util.json.parseJsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.add
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull

object SeriesWatchingNowSnapshotService {
    fun getSnapshot(accountId: String?, categoryId: String?, seriesId: String?): SeriesWatchingNowSnapshot? {
        if (StringUtils.isBlank(accountId) || StringUtils.isBlank(seriesId)) {
            return null
        }
        val normalizedCategory = normalize(categoryId)
        val candidateIds = buildSeriesIdCandidates(seriesId)
        val normalizedAccountId = accountId.orEmpty()
        val exact = findExactSnapshot(normalizedAccountId, normalizedCategory, candidateIds)
        return exact ?: findLatestSnapshot(normalizedAccountId, candidateIds)
    }
    fun loadEpisodeList(accountId: String?, categoryId: String?, seriesId: String?): EpisodeList {
        val snapshot = getSnapshot(accountId, categoryId, seriesId)
        if (snapshot == null || StringUtils.isBlank(snapshot.episodesJson)) {
            return EpisodeList()
        }
        val payload = parseJsonArray(snapshot.episodesJson.orEmpty()) ?: return EpisodeList()
        val list = EpisodeList()
        for (index in payload.indices) {
            val channel = readChannel(payload, index)
            if (channel != null) {
                list.episodes.add(toEpisode(channel))
            }
        }
        return list
    }
    fun loadChannels(accountId: String?, categoryId: String?, seriesId: String?): List<Channel> {
        val snapshot = getSnapshot(accountId, categoryId, seriesId)
        if (snapshot == null || StringUtils.isBlank(snapshot.episodesJson)) {
            return emptyList()
        }
        val payload = parseJsonArray(snapshot.episodesJson.orEmpty()) ?: return emptyList()
        val channels = ArrayList<Channel>()
        for (index in payload.indices) {
            val channel = readChannel(payload, index)
            if (channel != null) {
                channels.add(channel)
            }
        }
        return channels
    }
    fun save(
        account: Account?,
        categoryId: String?,
        seriesId: String?,
        categoryDbId: String?,
        seriesTitle: String?,
        seriesPoster: String?,
        episodeList: EpisodeList?
    ) {
        if (account == null || StringUtils.isBlank(account.dbId) || StringUtils.isBlank(seriesId) || episodeList?.episodes.isNullOrEmpty()) {
            return
        }
        val episodesPayload = buildJsonArray {
            episodeList.episodes.forEach { episode ->
                val channel = toChannel(episode)
                val payload = channel?.toJson()?.let(::parseJsonObject)
                if (payload != null) {
                    add(payload)
                }
            }
        }
        if (episodesPayload.isEmpty()) {
            return
        }
        val snapshot = SeriesWatchingNowSnapshot()
        snapshot.accountId = account.dbId
        snapshot.categoryId = normalize(categoryId)
        snapshot.seriesId = canonicalizeSeriesId(seriesId)
        snapshot.categoryDbId = normalize(categoryDbId)
        snapshot.seriesTitle = normalize(seriesTitle)
        snapshot.seriesPoster = normalize(seriesPoster)
        snapshot.episodesJson = episodesPayload.toString()
        snapshot.updatedAt = System.currentTimeMillis()
        SeriesWatchingNowSnapshotDb.get().upsert(snapshot)
    }
    fun saveChannels(
        account: Account?,
        categoryId: String?,
        seriesId: String?,
        categoryDbId: String?,
        seriesTitle: String?,
        seriesPoster: String?,
        channels: List<Channel>?
    ) {
        if (channels.isNullOrEmpty()) {
            return
        }
        val list = EpisodeList()
        channels.filterNotNull().forEach { list.episodes.add(toEpisode(it)) }
        save(account, categoryId, seriesId, categoryDbId, seriesTitle, seriesPoster, list)
    }
    fun clear(accountId: String?, categoryId: String?, seriesId: String?) {
        if (StringUtils.isBlank(accountId) || StringUtils.isBlank(seriesId)) {
            return
        }
        val canonicalSeriesId = canonicalizeSeriesId(seriesId)
        val normalizedCategory = normalize(categoryId)
        val normalizedAccountId = accountId.orEmpty()
        SeriesWatchingNowSnapshotDb.get().clear(normalizedAccountId, normalizedCategory, canonicalSeriesId)
        for (candidateId in buildSeriesIdCandidates(seriesId)) {
            for (candidate in SeriesWatchingNowSnapshotDb.get().getBySeries(normalizedAccountId, candidateId)) {
                SeriesWatchingNowSnapshotDb.get().clear(normalizedAccountId, normalize(candidate.categoryId), normalize(candidate.seriesId))
            }
        }
    }
    fun clearAll() {
        SeriesWatchingNowSnapshotDb.get().clearAll()
    }

    private fun toChannel(episode: Episode?): Channel? {
        if (episode == null || StringUtils.isBlank(episode.id)) {
            return null
        }
        val channel = Channel()
        channel.channelId = episode.id
        channel.name = episode.title
        channel.cmd = episode.cmd
        channel.season = episode.season
        channel.episodeNum = episode.episodeNum
        channel.extraJson = episode.toJson()
        episode.info?.let { info ->
            channel.logo = info.movieImage
            channel.description = info.plot
            channel.releaseDate = info.releaseDate
            channel.rating = info.rating
            channel.duration = info.duration
        }
        return channel
    }

    private fun toEpisode(channel: Channel): Episode {
        val parsed = Episode.fromJson(channel.extraJson)
        if (parsed != null && StringUtils.isNotBlank(parsed.id)) {
            hydrateParsedEpisode(parsed, channel)
            return parsed
        }
        val episode = Episode()
        episode.id = channel.channelId
        episode.title = channel.name
        episode.cmd = channel.cmd
        episode.season = channel.season
        episode.episodeNum = channel.episodeNum
        val info = EpisodeInfo()
        info.movieImage = channel.logo
        info.plot = channel.description
        info.releaseDate = channel.releaseDate
        info.rating = channel.rating
        info.duration = channel.duration
        episode.info = info
        return episode
    }

    private fun normalize(value: String?): String = value?.trim().orEmpty()

    private fun findExactSnapshot(accountId: String, normalizedCategory: String, candidateIds: List<String>): SeriesWatchingNowSnapshot? {
        for (candidateId in candidateIds) {
            val match = SeriesWatchingNowSnapshotDb.get().getBySeries(accountId, normalizedCategory, candidateId)
            if (match != null) {
                return match
            }
        }
        return null
    }

    private fun findLatestSnapshot(accountId: String, candidateIds: List<String>): SeriesWatchingNowSnapshot? {
        var latest: SeriesWatchingNowSnapshot? = null
        candidateIds.forEach { candidateId ->
            latest = newerSnapshot(latest, SeriesWatchingNowSnapshotDb.get().getBySeries(accountId, candidateId))
        }
        return latest
    }

    private fun newerSnapshot(current: SeriesWatchingNowSnapshot?, candidates: List<SeriesWatchingNowSnapshot>): SeriesWatchingNowSnapshot? {
        var latest = current
        candidates.forEach { candidate ->
            if (latest == null || candidate.updatedAt > latest.updatedAt) {
                latest = candidate
            }
        }
        return latest
    }

    private fun readChannel(payload: kotlinx.serialization.json.JsonArray, index: Int): Channel? =
        payload.getOrNull(index)?.let { element ->
            when (element) {
                is JsonPrimitive -> element.contentOrNull?.let(Channel::fromJson)
                else -> Channel.fromJson(element.toString())
            }
        }

    private fun hydrateParsedEpisode(parsed: Episode, channel: Channel) {
        if (StringUtils.isBlank(parsed.season)) {
            parsed.season = channel.season
        }
        if (StringUtils.isBlank(parsed.episodeNum)) {
            parsed.episodeNum = channel.episodeNum
        }
        if (parsed.info == null) {
            parsed.info = EpisodeInfo()
        }
        mergeEpisodeInfo(parsed.info, channel)
    }

    private fun mergeEpisodeInfo(info: EpisodeInfo?, channel: Channel) {
        if (info == null) {
            return
        }
        if (StringUtils.isBlank(info.movieImage)) {
            info.movieImage = channel.logo
        }
        if (StringUtils.isBlank(info.plot)) {
            info.plot = channel.description
        }
        if (StringUtils.isBlank(info.releaseDate)) {
            info.releaseDate = channel.releaseDate
        }
        if (StringUtils.isBlank(info.rating)) {
            info.rating = channel.rating
        }
        if (StringUtils.isBlank(info.duration)) {
            info.duration = channel.duration
        }
    }

    private fun canonicalizeSeriesId(seriesId: String?): String {
        val raw = normalize(seriesId)
        if (StringUtils.isBlank(raw) || !raw.contains(":")) {
            return raw
        }
        val last = raw.split(":").asReversed().firstOrNull { StringUtils.isNotBlank(it.trim()) }?.trim().orEmpty()
        return if (StringUtils.isBlank(last)) raw else last
    }

    private fun buildSeriesIdCandidates(seriesId: String?): List<String> {
        val raw = normalize(seriesId)
        if (StringUtils.isBlank(raw)) {
            return emptyList()
        }
        val candidates = LinkedHashSet<String>()
        candidates.add(canonicalizeSeriesId(raw))
        if (raw.contains(":")) {
            raw.split(":").forEach { part ->
                val normalized = normalize(part)
                if (StringUtils.isNotBlank(normalized)) {
                    candidates.add(normalized)
                }
            }
        } else {
            candidates.add("$raw:$raw")
        }
        candidates.add(raw)
        return ArrayList(candidates)
    }
}
